pkg/asset/manifests/aws/zones.go

```diff
@@ -0,0 +1,272 @@
+	if zin.InstallConfig == nil {
+		return fmt.Errorf("failed to get installConfig: %w", err)
+	}
+	if zin.InstallConfig.AWS == nil {
+		return fmt.Errorf("failed to get AWS metadata: %w", err)
+	}
+	zin.Subnets = &subnetsInput{}
+	if zin.Subnets.privateSubnets, err = zin.InstallConfig.AWS.PrivateSubnets(context.TODO()); err != nil {
+		return fmt.Errorf("failed to get private subnets: %w", err)
+	}
+	if zin.Subnets.publicSubnets, err = zin.InstallConfig.AWS.PublicSubnets(context.TODO()); err != nil {
+		return fmt.Errorf("failed to get public subnets: %w", err)
+	}
+	if zin.Subnets.vpc, err = zin.InstallConfig.AWS.VPC(context.TODO()); err != nil {
+		return fmt.Errorf("failed to get VPC: %w", err)
+	}
+	return nil
+}
+
+type zonesCAPI struct {
+	AvailabilityZones sets.Set[string]
+	controlPlaneZones sets.Set[string]
+	computeZones      sets.Set[string]
+}
+
+// SetAvailabilityZones insert the zone to the given compute pool, and to
+// the regular zone (zone type availability-zone) list.
+func (zo *zonesCAPI) SetAvailabilityZones(pool string, zones []string) {
+	switch pool {
+	case types.MachinePoolControlPlaneRoleName:
+		zo.controlPlaneZones.Insert(zones...)
+
+	case types.MachinePoolComputeRoleName:
+		zo.computeZones.Insert(zones...)
+	}
+	zo.AvailabilityZones.Insert(zones...)
+}
+
+// SetDefaultConfigZones evaluates if machine pools (control plane and workers) have been
+// set the zones from install-config.yaml, if not sets the default from platform, when exists,
+// otherwise set the default from the region discovered from AWS API.
+func (zo *zonesCAPI) SetDefaultConfigZones(pool string, defConfig []string, defRegion []string) {
+	zones := []string{}
+	switch pool {
+	case types.MachinePoolControlPlaneRoleName:
+		if len(zo.controlPlaneZones) == 0 && len(defConfig) > 0 {
+			zones = defConfig
+		} else if len(zo.controlPlaneZones) == 0 {
+			zones = defRegion
+		}
+		zo.controlPlaneZones.Insert(zones...)
+
+	case types.MachinePoolComputeRoleName:
+		if len(zo.computeZones) == 0 && len(defConfig) > 0 {
+			zones = defConfig
+		} else if len(zo.computeZones) == 0 {
+			zones = defRegion
+		}
+		zo.computeZones.Insert(zones...)
+	}
+	zo.AvailabilityZones.Insert(zones...)
+}
+
+// setZones is the entrypoint to create the CAPI NetworkSpec structures
+// for managed or BYO VPC deployments from install-config.yaml.
+func setZones(in *zonesInput) error {
+	if len(in.InstallConfig.Config.AWS.Subnets) > 0 {
+		if err := in.GatherSubnetsFromMetadata(); err != nil {
+			return fmt.Errorf("failed to get subnets from metadata: %w", err)
+		}
+		return setZonesBYOVPC(in)
+	}
+
+	if err := in.GatherZonesFromMetadata(); err != nil {
+		return fmt.Errorf("failed to get availability zones from metadata: %w", err)
+	}
+	return setZonesManagedVPC(in)
+}
+
+// setZonesBYOVPC creates the CAPI NetworkSpec.Subnets setting the
+// desired subnets from install-config.yaml in the BYO VPC deployment.
+func setZonesBYOVPC(in *zonesInput) error {
```

Comment:
Nit:
```suggestion
func setZonesBYOVPC(ctx context.Context, in *zonesInput) error {
```
If you move the zone gathering here